### PR TITLE
[BEAM-957] Use a Multiset to track Pending Bundles

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

A bundle may be consumed by a Flatten multiple times. The last instance
completing should remove the hold caused by the bundle, not the first.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles 

WIP: Unit test for WatermarkManager, plus I'd like an update to `FlattenTest` to make this output more readable.

**Comment by tgroh** [comment]
New tests are done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlattenTest.java [hunk 1]
```diff
@@ -121,16 +121,30 @@ public void testFlattenPCollectionListEmpty() {
...
-    PCollection<String> flattened = PCollectionList.of(lines)
-        .and(lines2)
-        .and(lines)
-        .and(lines)
-        .apply(Flatten.<String>pCollections());
-
-    PAssert.that(flattened).containsInAnyOrder(Iterables.concat(LINES, LINES, LINES, LINES2));
+    PCollection<Long> longs = p.apply("mkLines", CountingInput.upTo(10));
+    PCollection<Long> biggerLongs =
+        p.apply("mkOtherLines", CountingInput.upTo(10))
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlattenTest.java]: No action required, but having a starting point seems nicer than mapping over it. This test is a demonstration that it is useful. General goal: multiple PCollections automatically generated with disjoint elements.

**Comment by kennknowles** [comment]
Filed [BEAM-961](https://issues.apache.org/jira/browse/BEAM-961) as a nice little starter task for my comment.

**Pull request closed without merging**